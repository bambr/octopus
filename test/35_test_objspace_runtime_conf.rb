#!/usr/bin/ruby

$: << File.dirname($0) + '/lib'
require 'run_env'

class Env < RunEnv
end

env = Env.new

env.connect_eval do
  insert ["foo"]
  env.snapshot
  insert ["bar"]

  env.stop
end

env.connect_eval do
  select "foo", "bar"
  env.stop
end

env.cd do
  File.open(Env::ConfigFile, "w") { |io| io.puts(env.config :object_space => false) }
end

env.env_eval do
  cd do
    invoke :start
    env.octopus [], :out => "/dev/null", :err => "/dev/null"
    err = ""
    wait_for "server failure" do
      err = File.read(Env::LogFile).match(/exception: (shard|object_space) 0 is not configured/)
    end
    puts err
    puts
    rm Env::PidFile
  end
end

env.cd do
  File.open(Env::ConfigFile, "w") { |io| io.puts(env.config) }
end

env.connect_eval do
  select "foo", "bar"
  insert ["baz"]
  log_try { create_index 1, :type => :FASTTREE, :unique => 1, :field_0 => { :type => :STRING, :index => 0 , :sort_order => :DESC } }
  env.stop
end

env.cd do
  File.open(Env::ConfigFile, "w") { |io| io.puts(env.config :hostname => "one") }
end

env.connect_eval do
  create_shard 0, :POR, "one"
  sleep 0.2
  # create_object_space 0, :shard => 0, :index => {:type => :FASTTREE, :unique => 1, :field_0 => { :type => :STRING, :index => 0 , :sort_order => :DESC }}
  create_index 1, :type => :FASTTREE, :unique => 1, :field_0 => { :type => :STRING, :index => 0 , :sort_order => :DESC }

  create_shard 1, :POR, "one"
  create_object_space 0, :shard => 1, :index => {:type => :FASTTREE, :unique => 1, :field_0 => { :type => :STRING, :index => 0 , :sort_order => :DESC }}

  env.snapshot
  insert ["baf"]
  env.stop
end

env.connect_eval do
  select "foo", "bar", "baz", "baf"
  env.stop
end

env.cd do
  puts `./octopus --cat 00000000000000000008.snap | sed 's/tm:[0-9.]\\+ //g;'`
end
