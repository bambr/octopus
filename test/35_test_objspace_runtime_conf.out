# box.insert(["foo"])
1

# box.insert(["bar"])
1

# box.select("foo", "bar")
[["foo"], ["bar"]]

F> exception: object_space 0 is not configured

# box.select("foo", "bar")
[["foo"], ["bar"]]

# box.insert(["baz"])
1

# box.create_index(1, {:type=>:FASTTREE, :unique=>1, :field_0=>{:type=>:STRING, :index=>0, :sort_order=>:DESC}})
Failed with: {code: 0x202, message: 'metadata updates are forbidden in legacy configuration mode'}
reloading config
ok

# box.create_index(1, {:type=>:FASTTREE, :unique=>1, :field_0=>{:type=>:STRING, :index=>0, :sort_order=>:DESC}})
""

# box.insert(["baf"])
1

# box.select("foo", "bar", "baz", "baf")
[["foo"], ["bar"], ["baz"], ["baf"]]

lsn:5 scn:5 t:sys/snap_initial 0.0.0.0:0 count:3 run_crc_log:0x3e239886 run_crc_mod:0x00000000
lsn:5 scn:5 t:snap/usr240 0.0.0.0:0 CREATE_OBJECT_SPACE n:0 flags:00000000 cardinalty:-1 snap:1 wal:1 PK: min_tuple_cardinality:1 cardinality:1 type:0 unique:1 field0:{index:0 type:7 sort:1}
lsn:5 scn:5 t:snap/snap_data 0.0.0.0:0 n:0 <"bar">
lsn:5 scn:5 t:snap/snap_data 0.0.0.0:0 n:0 <"baz">
lsn:5 scn:5 t:snap/snap_data 0.0.0.0:0 n:0 <"foo">
lsn:5 scn:5 t:snap/usr241 0.0.0.0:0 CREATE_INDEX n:0 flags:00000000 min_tuple_cardinality:1 cardinality:1 type:4 unique:1 field0:{index:0 type:7 sort:-1}
lsn:5 scn:5 t:sys/snap_final 0.0.0.0:0 
